rust/lance-core/src/utils/tokio.rs

```diff
@@ -1,16 +1,35 @@
 // SPDX-License-Identifier: Apache-2.0
 // SPDX-FileCopyrightText: Copyright The Lance Authors
 
+use std::time::Duration;
+
 use crate::Result;
 
 use futures::{Future, FutureExt};
 use tokio::runtime::{Builder, Runtime};
 use tracing::Span;
 
+fn get_num_compute_intensive_cpus() -> usize {
+    if num_cpus::get() <= *IO_CORE_RESERVATION {
+        log::warn!(
+            "Number of CPUs is less than or equal to the number of IO core reservations. \
+            This is not a supported configuration. using 1 CPU for compute intensive tasks."
+        );
+        return 1;
+    }
+
+    num_cpus::get() - *IO_CORE_RESERVATION
```

Comment:
We should re-use the `num_cpus::get()` call instead of making a sys call repeatedly. Also, I think if there's only 1 CPU, maybe we shouldn't bother with the warning.
```suggestion
    let cpu_count = num_cpus::get();
    if cpu_count <= *IO_CORE_RESERVATION && cpu_count > 1 {
        log::warn!(
            "Number of CPUs is less than or equal to the number of IO core reservations. \
            This is not a supported configuration. using 1 CPU for compute intensive tasks."
        );
        return 1;
    }

    cpu_count - *IO_CORE_RESERVATION
```

Reply:
I moved the `cpu_count > 1` to inside the block, because we still want to return 1 when there is only one CPU.